packages/arb-token-bridge-ui/src/hooks/useRedeemTeleporter.ts

```diff
@@ -0,0 +1,226 @@
+const redeemTeleporterSecondLeg = async ({
+  tx,
+  signer,
+  txUpdateCallback
+}: {
+  tx: MergedTransaction
+  signer: Signer
+  txUpdateCallback?: (tx: MergedTransaction) => Promise<void>
+}) => {
+  // check if we require a redemption for the l2l3 retryable
+  if (
+    secondRetryableLegForTeleportRequiresRedeem(tx) &&
+    tx.l1ToL2MsgData?.l2TxID &&
+    tx.l2ToL3MsgData
+  ) {
+    const l2L3Retryable = await getRetryableTicket({
+      parentChainTxHash: tx.l1ToL2MsgData.l2TxID,
+      retryableCreationId: tx.l2ToL3MsgData?.retryableCreationTxID,
+      parentChainProvider: getProviderForChainId(tx.l2ToL3MsgData.l2ChainId),
+      childChainSigner: signer
+    })
+
+    // redeem retryable
+    const redemptionReceipt = await redeemRetryable(l2L3Retryable)
+
+    // update the teleport tx in the UI
+    const l2ToL3MsgData: L2ToL3MessageData = {
+      ...tx.l2ToL3MsgData,
+      l3TxID: redemptionReceipt.l2TxReceipt.transactionHash,
+      status: L1ToL2MessageStatus.REDEEMED,
+      retryableCreationTxID: l2L3Retryable.retryableCreationId
+    }
+    const updatedTeleporterTx = await getUpdatedTeleportTransfer({
+      ...tx,
+      l2ToL3MsgData,
+      depositStatus: DepositStatus.L2_SUCCESS,
+      resolvedAt: dayjs().valueOf()
+    })
+    await txUpdateCallback?.(updatedTeleporterTx)
+  }
+}
+
+const getUpdatedTeleporterTxAfterRedemption = async (tx: MergedTransaction) => {
+  const { l1ToL2MsgData, l2ToL3MsgData } =
+    await fetchTeleporterDepositStatusData(tx)
+
+  const teleportTransfer = {
+    ...tx,
+    l1ToL2MsgData,
+    l2ToL3MsgData
+  }
+
+  return {
+    ...teleportTransfer,
+    depositStatus: getDepositStatus(teleportTransfer)
+  }
+}
+
+export function useRedeemTeleporter(
+  tx: MergedTransaction,
+  address: Address | undefined
+): UseRedeemRetryableResult {
+  const chainIdForRedeemingRetryable = getChainIdForRedeemingRetryable(tx)
+
+  const { data: signer } = useSigner({
+    chainId: chainIdForRedeemingRetryable
+  })
+  const { updatePendingTransaction } = useTransactionHistory(address)
+
+  const redeemerNetworkName = getNetworkName(chainIdForRedeemingRetryable)
+
+  const [isRedeeming, setIsRedeeming] = useState(false)
+
+  const redeem = useCallback(async () => {
+    if (isRedeeming) {
+      return
+    }
+
+    if (!isTeleport(tx)) {
+      throw new Error(
+        'The transaction being redeemed is not a LayerLeap transaction.'
+      )
+    }
+
+    try {
+      setIsRedeeming(true)
+
+      if (!signer) {
+        throw 'Signer is undefined'
+      }
+
+      if (firstRetryableLegRequiresRedeem(tx)) {
+        await redeemTeleporterFirstLeg({
+          tx,
+          signer,
+          txUpdateCallback: updatePendingTransaction
+        })
+      } else if (secondRetryableLegForTeleportRequiresRedeem(tx)) {
+        await redeemTeleporterSecondLeg({
+          tx,
+          signer,
+          txUpdateCallback: updatePendingTransaction
+        })
+      }
+
+      // track in analytics
```

Comment:
What if both `firstRetryableLegRequiresRedeem` and `secondRetryableLegForTeleportRequiresRedeem` are false?